### PR TITLE
Do not charge the idle sc

### PR DESCRIPTION
**Pull request by michaelmcinerney**
Anna mentioned in https://github.com/seL4/seL4/issues/684 the idea that the idle sc could have infinite budget, and so would always be ready and sufficient. I did try this out, but it was quite difficult to get things working, and I have abandoned this idea for now. Perhaps we can go with the approach in this PR now, and reconsider this infinite budget idea later down the track

### CODE REVIEW

**Comment by michaelmcinerney** [comment]
This has been rebased and includes the update for the Haskell and Refine. I really tried to update `chargeBudget_corres` in as minimal a way as possible, but I found it very difficult. If anyone has an idea for  how to do that better, that would be great

**Comment by michaelmcinerney** [comment]
I didn't realise the file `Init_R.thy` had been added to the ROOT file for Refine. Hopefully the proofs will pass now (I think everything else passed the previous attempt)

**Comment by michaelmcinerney** [comment]
For convenience, here's Kent's draft PR implementing these changes in the C

https://github.com/seL4/seL4/pull/697

File: proof/invariant-abstract/DetSchedSchedule_AI.thy [hunk 1]
```diff
@@ -21772,8 +21806,40 @@ lemma reset_untyped_cap_cur_sc_active_implies_cur_sc_offset_ready[wp]:
...
+    apply wpsimp
+   apply wpsimp
+  apply (rule hoare_seq_ext_skipE)
+   apply (subst cur_sc_active_implies_cur_sc_offset_ready_rewrite)+
+   apply (rule hoare_validE_conj)
+    apply (rule valid_validE)
+    apply (rule hoare_weaken_pre)
+     apply (rule hoare_vcg_imp_lift')
+      apply wpsimp
+     apply wpsimp
```
[the remark this answers is not in the document]
Comment (mktnk3) [reply proof/invariant-abstract/DetSchedSchedule_AI.thy]: Looks like this is a recurring pattern in other places, too, maybe? Not sure if there is a way to package this up.

**Review by mktnk3 — APPROVED** [review]
Nice work. Great to see this working well!

Maybe this is a bit late, but it might help to have an abbreviation for `\<lambda>s. cur_sc s \<noteq> idle_sc_ptr`.

File: proof/invariant-abstract/DetSchedInvs_AI.thy [hunk 2]
```diff
@@ -3901,6 +3903,32 @@ lemma cur_sc_active_ct_not_in_release_q_imp_ct_running_imp_ct_schedulable:
...
+  "\<lbrakk>valid_idle s; sym_refs (state_refs_of s); ct_active s; cur_sc_tcb_are_bound s\<rbrakk>
+   \<Longrightarrow> cur_sc s \<noteq> idle_sc_ptr"
+  apply (clarsimp simp: valid_idle_def vs_all_heap_simps ct_in_state_def pred_tcb_at_def obj_at_def)
+  apply (frule_tac tp="cur_thread s" in sym_ref_tcb_sc; fastforce?)
+  apply fastforce
+  done
+
+lemma cur_sc_not_idle_sc_ptr':
+  "\<lbrakk>invs s; ct_running s; cur_sc_tcb_are_bound s\<rbrakk>
+   \<Longrightarrow> cur_sc s \<noteq> idle_sc_ptr"
```
[the remark this answers is not in the document]
Comment (michaelmcinerney) [reply proof/invariant-abstract/DetSchedInvs_AI.thy:3926]: The second version is used in the Eisbach method `he_ctris_ff`. I couldn't get the first version working there. Those Eisbach methods are a real pain sometimes, so I thought I'd just stick with this. 

As for the name, I can't think up one that's much better, but if anyone has any ideas, I'd be interested

File: proof/invariant-abstract/DetSchedSchedule_AI.thy (hunk 1 above)
[the remark this answers is not in the document]
Comment (michaelmcinerney) [reply proof/invariant-abstract/DetSchedSchedule_AI.thy]: Yeah something along those lines happens here and in a few other proofs near here. I tried very hard to make these nicer, but from memory, there isn't one pattern that recurs every time, and I think I got into a lot of trouble if I tried to use the pipes and plusses with these rules

**Review by corlewis — APPROVED** [review]
Great job pushing this through with the related C change!

**Pull request merged**